### PR TITLE
Add model registration argument for autolog

### LINKED ISSUE #4829: [FR] Simplify model registration in autolog
## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [X] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary

`autolog()` makes tracking a lot simpler but it does not seem to provide an option to register models. That leads to code like this where the user needs to format a `runs:/` URL to be able to register the model, which is pretty user-unfriendly. Ideally, model registration would also not require special-casing `file:` during local runs.

```python
import mlflow
import mlflow.sklearn
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

# Setup MLFlow tracking
mlflow.sklearn.autolog()

# Load data
X, y = load_diabetes(return_X_y = True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

with mlflow.start_run() as run:
    # Train model
    model = Ridge(alpha=0.03)
    model.fit(X_train, y_train)

    # Test model
    mse = mean_squared_error(model.predict(X_test), y_test)
    mlflow.log_metric("test_mse", mse)

    # Register model
    if not(mlflow.get_tracking_uri().startswith('file:')):
        mlflow.register_model(f"runs:/{run.info.run_id}/model", "my_model")
```

## Motivation
- What is the use case for this feature?

Simplify model registration.

- Why is this use case valuable to support for MLflow users in general?

This creates a barrier to entry for new users: I had to spend quite some time figuring out the right string-format incantation.

- Why is this use case valuable to support for your project(s) or organization?

I am trying to evangelize more rigorous MLOps processes and extra complexity reduces appeal.

- Why is it currently difficult to achieve this use case? (please be as specific as possible about why related MLflow features and components are insufficient)

String formatting is very low level and leads to typos difficult to debug.

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

## Details

Implementation-wise, one possibility would be for `mlflow.register_model()` to accept a `run` object as input, or for `autolog()` to give an option to register models.

Note sure how to deal with the `file:` special-case. Maybe log instead of failing (or implement model registration in `file:` but that's likely larger work).

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## What changes are proposed in this pull request?

Add model registration argument for autolog

## How is this patch tested?

Unit tests.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add model registration argument for autolog

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/lightgbm.py
Comment (harupy) [inline mlflow/lightgbm.py:620]:
```suggestion
                FLAVOR_NAME, "registered_model_name", None
```

Comment (harupy) [reply mlflow/lightgbm.py:620]: same comment for other flavors

File: tests/lightgbm/test_lightgbm_autolog.py
Comment (harupy) [inline tests/lightgbm/test_lightgbm_autolog.py:664]: why not check the model is registered with the specified name?

Comment (WeichenXu123) [reply tests/lightgbm/test_lightgbm_autolog.py:664]: that might requires more code to write, and `register_model` functionality should be covered by the model registry testing code.

Comment (harupy) [reply tests/lightgbm/test_lightgbm_autolog.py:664]: > that might require more code to write.

Are you sure? Can we just call `get_registered_model`? Let me know if I'm missing something.

Comment (harupy) [reply tests/lightgbm/test_lightgbm_autolog.py:664]: > `register_model` functionality should be covered by the model registry testing code.

Yes, but how do we ensure we don't pass incorrect arguments?

File: mlflow/fastai/__init__.py
Comment (dbczumar) [inline mlflow/fastai/__init__.py:409]: Can we update the docs as follows for each flavor?

```suggestion
    :param registered_model_name: If given, each time a model is trained, it is registered as a new
                                  model version of the registered model with this name. The registered model is
                                  created if it does not already exist.
```

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/5395/files#r813665077 is addressed. Thanks @WeichenXu123 !

File: tests/lightgbm/test_lightgbm_autolog.py
Comment (WeichenXu123) [reply tests/lightgbm/test_lightgbm_autolog.py]: updated using `MlflowClient().get_registered_model(name)` for test

**Review by harupy — APPROVED** [review]
LGTM once the lint check failure is fixed

**Comment by WeichenXu123** [comment]
A weird error in CI tensorflow 2.0.0 / 2.0.4
https://github.com/mlflow/mlflow/runs/5319733965?check_suite_focus=true#step:12:80

but I cannot reproduce it in my local machine. might be CI env setup issue.
shall this block the PR ?

**Comment by harupy** [comment]
> shall this block the PR ?

I think it does because the failed test (`test_tf_keras_model_autolog_registering_model`) was added by this PR.

**Comment by harupy** [comment]
@WeichenXu123 `clear_tf_keras_imports` might be the root cause. We could try running `test_tf_keras_model_autolog_registering_model` before `clear_tf_keras_imports`.

**Pull request merged**
